src/ot1d.jl

```diff
@@ -0,0 +1,283 @@
+  wj = v_w_sorted[1]
+  i,j = 1,1
+  while true
+      if (wi < wj || j == length(v_w_sorted))
+          γ[perm_u[i],perm_v[j]] = wi
+          i+=1
+          if i == length(u_sorted)+1
+              break
+          end
+          wj -= wi
+          wi  = u_w_sorted[i]
+      else
+          γ[perm_u[i],perm_v[j]] = wj
+          j+=1
+          if j == length(v_sorted)+1
+              break
+          end
+          wi -= wj
+          wj  = v_w_sorted[j]
+      end
+  end
+  
+  return γ
+end
+
+"""
+otCostPlan1d(c,u::Vector,u_weights::Vector,v::Vector,v_weights::Vector)
+otCostPlan1d(c,μ::Distributions.DiscreteNonParametric,ν::Distributions.DiscreteNonParametric)
+
+Calculates the Optimal Transport Cost between μ to ν, where
+they are 1-Dimensional distributions and the cost
+function is of the form ``c(x,y) = h(|x-y|)`` such that
+``h`` is a convex function.
+
+Note that when you declar `μ = DiscreteNonParametric(u,u_n)`, the
+distribution is sorted, hence, `otPlan1d(c,u,u_n,v,v_n)` will
+not be equal to `otPlan1d(c,μ,ν)` is `u` and/or `v` are not sorted.
+
+For the first case we have parameters:\n
+c: The cost function, which should be of the form ``c(x,y) = h(abs(x-y))``
+where ``h`` is a convex function. \n
+u        : Vector of point mass location    (e.g. `u = collect(1:10)`)\n
+u_weights: Vector of weights for each point (e.g. `u_weights = ones(10)*0.1`)\n
+v        : Vector of point mass location    (e.g. `v = collect(10:20)`)\n
+v_weights: Vector of weights for each point (e.g. `v_weights = ones(10)*0.1`)\n
+
+
+For the second case we have parameters:\n
+c: The cost function, which should be of the form ``c(x,y) = h(abs(x-y))``
+where ``h`` is a convex function. \n
+μ: Finite Discrete Distribution (e.g. `μ = DiscreteNonParametric(u, u_n)`)\n
+ν: Finite Discrete Distribution (e.g. `ν = DiscreteNonParametric(v, v_n)`)\n
+
+Returns the Optimal Transport Cost and the Plan as a matrix γ.
+"""
+function otCostPlan1d(c,u::Vector,u_weights::Vector,v::Vector,v_weights::Vector)
+  γ = zeros(length(u),length(v))
+  cost = 0
+  perm_u = sortperm(u)
+  perm_v = sortperm(v)
+  u_sorted = u[perm_u]
+  v_sorted = v[perm_v]
+  u_w_sorted = u_weights[perm_u]
+  v_w_sorted = v_weights[perm_v]
+  
+  wi = u_w_sorted[1]
+  wj = v_w_sorted[1]
+  i,j = 1,1
+  while true
+      if (wi < wj || j == length(v_w_sorted))
+          cost += c(u_sorted[i],v_sorted[j])*wi
+          γ[perm_u[i],perm_v[j]] = wi
+          i+=1
+          if i == length(u_sorted)+1
+              break
+          end
+          wj -= wi
+          wi  = u_w_sorted[i]
+      else
+          γ[perm_u[i],perm_v[j]] = wj
+          cost += c(u_sorted[i],v_sorted[j])*wj
+          j+=1
+          if j == length(v_sorted)+1
+              break
+          end
+          wi -= wj
+          wj  = v_w_sorted[j]
+      end
+  end
+  
+  return cost, γ
+end
+
+function otCost1d(c,μ::Distributions.DiscreteNonParametric,
+                    ν::Distributions.DiscreteNonParametric)
```

Comment:
As mentioned, I think one should *only* accept DiscreteNonParametric, no vectors.

Reply:
I don't have objections to this. If you think it's better, than we can go with that.